test/infrastructure/docker/exp/internal/docker/nodepool.go

```diff
@@ -56,16 +57,48 @@ type NodePool struct {
 	dockerMachinePool *infraexpv1.DockerMachinePool
 	labelFilters      map[string]string
 	machines          []*docker.Machine
+	nodePoolInstances []NodePoolInstance // Note: This must be initialized when creating a new node pool and updated to reflect the `machines` slice.
+}
+
+// NodePoolInstance is a representation of a node pool instance used to provide the information to construct DockerMachines.
+type NodePoolInstance struct {
+	InstanceName     string
+	Bootstrapped     bool
+	ProviderID       *string
+	PrioritizeDelete bool
+	Addresses        []clusterv1.MachineAddress
+	Ready            bool
+}
+
+func (np NodePool) Len() int      { return len(np.machines) }
+func (np NodePool) Swap(i, j int) { np.machines[i], np.machines[j] = np.machines[j], np.machines[i] }
+func (np NodePool) Less(i, j int) bool {
```

Comment:
I think it makes sense to define a new type that's sortable. However, it seems a bit odd to have a type `NodePoolMachines` representing a `[]*docker.Machine` and a separate type `NodePoolMachine` representing the instance information. Since the `NodePool` struct will look like this:

```
type NodePool struct {
	client            client.Client
	cluster           *clusterv1.Cluster
	machinePool       *expv1.MachinePool
	dockerMachinePool *infraexpv1.DockerMachinePool
	labelFilters      map[string]string
	machines          NodePoolMachines
	nodePoolMachines  []NodePoolMachine // Note: This must be initialized when creating a new node pool and updated to reflect the `machines` slice.
}
```

Maybe we can change the type names so they don't overlap as much?

Reply:
I've refactored it to be like this where we're using a wrapper struct instead of two parallel lists:

```
// NodePoolMachine is a wrapper around a docker.Machine and a NodePoolMachineStatus, which maintains additional information about the machine.
// At least one of Machine or Status must be non-nil.
type NodePoolMachine struct {
	Machine *docker.Machine
	Status  *NodePoolMachineStatus
}

// NodePoolMachineStatus is a representation of the additional information needed to reconcile a docker.Machine.
// This is the only source of truth for understanding if the machine is already bootstrapped, ready, etc. so this
// information must be maintained through reconciliation.
type NodePoolMachineStatus struct {
	Name             string
	PrioritizeDelete bool
}
```